### PR TITLE
@mzikherman - Adds is_saved field to artwork schema

### PR DESCRIPTION
**Pull request by broskoski**
Like the title says.

~~This actually _doesn't_ do the fancy batching that we want it to do right now as it's returning a new `DataLoader` every time. I'm still trying to figure out a nice way of passing the required options down to make the request, but wanted to get this in front of some eyes in the meantime.~~

Edit: Spoke too soon, I figured it out.

### CODE REVIEW

File: lib/loaders/saved_artwork.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  return new DataLoader(ids => {
+    return gravity.with(accessToken)('collection/saved-artwork/artworks', {
+      user_id: userId,
+      artworks: ids,
+      private: true,
+    }).then((results) => {
+      const parsedResults = map(ids, (id) => {
+        const match = find(results, { id });
+        if (match) return extend(match, { is_saved: true });
+        return { id, is_saved: false };
```
Comment (broskoski) [inline lib/loaders/saved_artwork.js]: DataLoader requires that you send back an array of the same size to resolve the promise, but gravity currently just doesn't include the artwork in the response if the artwork isn't saved.

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Please assign someone to merge this PR, and optionally include people who should review.</td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: schema/artwork/index.js [hunk 2]
```diff
@@ -407,6 +408,16 @@ export const artworkFields = () => {
...
       },
     },
+    is_saved: {
+      type: GraphQLBoolean,
+      resolve: ({ id }, {}, request, { rootValue: {
+        accessToken, userID } }) => {
+        if (!accessToken) return false;
+        return savedArtworkLoader
+          .load(id, { userID, accessToken })
+          .then(({ is_saved }) => is_saved);
```
Comment (broskoski) [inline schema/artwork/index.js]: Pretty straightforward here.

File: schema/artwork/index.js [hunk 3]
```diff
@@ -407,6 +408,16 @@ export const artworkFields = () => {
         );
       },
     },
+    is_saved: {
+      type: GraphQLBoolean,
+      resolve: ({ id }, {}, request, { rootValue: {
+        accessToken, userID } }) => {
+        if (!accessToken) return false;
+        return savedArtworkLoader(userID, accessToken)
```
Comment (broskoski) [inline schema/artwork/index.js]: Right so here we are new-ing up a `DataLoader`, and then calling the load. I suppose the options can be passed and parsed with as the key...

File: lib/loaders/saved_artwork.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import gravity from './gravity';
+import {
+  map,
+  find,
+  extend,
+  first,
+} from 'lodash';
+
+const savedArtworkLoader = new DataLoader(keys => {
+  const keyObjs = map(keys, JSON.parse);
```
Comment (broskoski) [inline lib/loaders/saved_artwork.js]: That's right, stringify-ing the options we need.

**Comment by broskoski** [comment]
screenshot 2017-04-12 19 31 08

screenshot 2017-04-12 19 30 42

**Comment by mzikherman** [comment]
A lil' spec would be cool but that looks awesome!!! 🎉 🌮 

cc @alloy our first batched data loader

**Comment by alloy** [comment]
Yeehaw! 🤠

we shouldn't be making any new singleton DataLoader instances, though. Instead, how about making a new instance that already holds the authentication token and inject it at the start of the request like the other new DataLoaders?

In addition, I don't think you need to use the `gravity` DataLoader singleton, but instead use the underlying http module?

**Comment by broskoski** [comment]
@alloy not sure I understand any of that! 👀 

Which new DataLoaders don't new themselves up? And why wouldn't we use the `gravity` DataLoader?

**Comment by broskoski** [comment]
@mzikherman and @alloy added a spec and switched the gravity loaded to be the wrapped fetch call instead of using the `gravityLoader`. I'm still not totally sure how or why to not make the `savedArtwork` loader work how it is currently. I find the per_type loader to be a little less intuitive than just making a new DataLoader in the way the documentation recommends. Maybe you can give me a little pseudo-code example of what you're thinking?

**Comment by alloy** [comment]
Sure thing, will do so tonight! Can you also link to what part of the documentation you're referring?

**Comment by broskoski** [comment]
@alloy sure! https://github.com/facebook/dataloader/blob/master/examples/SQL.md

**Comment by alloy** [comment]
Hmm, not sure I get from that example what the part is that you’re referring to. The per-request/per-type pattern is shown in the docs here: https://github.com/facebook/dataloader#caching-per-request

**Comment by alloy** [comment]
I started refactoring this but it quickly changed into a much larger refactor, so by this point I’d say that this can imo be merged and my refactor PR comes later.

File: test/lib/loaders/saved_artwork.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          { id: 'cab', title: "Queen's Ship", artists: [] },
+        ],
+      }));
+
+    savedArtwork.__Rewire__('gravity', gravity);
+
+    return savedArtwork.load(JSON.stringify({ id, userID, accessToken }))
+      .then(artwork => {
+        expect(gravity.args[0][0])
+          .toBe('collection/saved-artwork/artworks?artworks%5B%5D=cab&private=true&user_id=fake');
```
Comment (mzikherman) [inline test/lib/loaders/saved_artwork.js]: nice spec

**Comment by mzikherman** [comment]
Awesome, rebase + merge at will!

**Pull request merged**